### PR TITLE
Add PUT metadata end point

### PR DESCRIPTION
**Pull request by jondewijones**
### What

- Added PUT metadata end point
- Used `dphandlers` package for getting auth details on GetAll handler

Note: the new PUT endpoint has been added under `dataset` directory rather than `handlers` directory as that was the original place. It will tidied up in a following task/PR.

### Who can review

Anyone

### CODE REVIEW

File: dataset/put-metadata.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+func PutMetadata(dc DatasetClient, zc ZebedeeClient) http.HandlerFunc {
+	return dphandlers.ControllerHandler(func(w http.ResponseWriter, r *http.Request, lang, collectionID, accessToken string) {
+		putMetadata(w, r, dc, zc, accessToken, collectionID, lang)
+	})
+}
+
+func putMetadata(w http.ResponseWriter, req *http.Request, dc DatasetClient, zc ZebedeeClient, userAccessToken, collectionID, lang string) {
+	ctx := req.Context()
+
+	if userAccessToken == "" {
```
Comment (rav-pradhan) [inline dataset/put-metadata.go]: It looks like this is the same logic for the `getMetaData` handler function; could it be worthwhile extracting these two checks into a helper function(s)?

File: dataset/put-metadata.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+		http.Error(w, "error adding dataset to collection", http.StatusInternalServerError)
+		return
+	}
+
+	err = zc.PutDatasetVersionInCollection(ctx, userAccessToken, "", collectionID, datasetID, edition, version, body.CollectionState)
+	if err != nil {
+		log.Event(ctx, "error adding version to collection", log.ERROR, log.Error(err), log.Data(logInfo))
+		http.Error(w, "error adding version to collection", http.StatusInternalServerError)
+	}
+
```
Comment (rav-pradhan) [inline dataset/put-metadata.go:89]: Should we be setting a `200` or `204` status code in the responsewriter here?

**Review by rav-pradhan — COMMENTED** [review]
Couple small points of discussion but generally looks fine. Presuming you'll be merging in [the other PR](https://github.com/ONSdigital/dp-publishing-dataset-controller/pull/21) as well, so will wait for that

File: dataset/put-metadata.go (hunk 2 above)
Comment (jondewijones) [reply dataset/put-metadata.go:89]: 👍

**Review by rav-pradhan — APPROVED** [review]

**Pull request merged**